### PR TITLE
ptl-tool: make more configurable and usable

### PR DESCRIPTION
**Pull request by batrick**
Make the script executable, add arguments/usage, and a basic README.

Signed-off-by: Patrick Donnelly <pdonnell@redhat.com>

### CODE REVIEW

**Comment by batrick** [comment]
@yuriw hope this helps!

**Comment by joscollin** [comment]
retest this please.

@yuriw I will try this out today, as it was late night yesterday.

File: src/script/ptl-tool.py [hunk 1]
```diff
@@ -1,27 +1,84 @@
...
+# this script:
+#
+#     PTL_TOOL_BASE_PATH=refs/remotes/upstream/
+#
+#
+# ** Here are some basic exmples to get started: **
+#
+# Merging PR #1234567 and #2345678 into a new test branch with a testing label added to the PR:
+#
+# $ env PTL_TOOL_BASE_PATH=refs/remotes/upstream/ src/scripts/ptl-tool.py --base master 1234567 2345678
```
[the remark this answers is not in the document]
Comment (joscollin) [reply src/script/ptl-tool.py]: The script is always expecting a remote named 'upstream', when my remote is 'origin git@github.com:ceph/ceph'
```
jcollin@smithi042:~/test$ env PTL_TOOL_BASE_PATH=refs/remotes/upstream/ src/script/ptl-tool.py --base master 17968
Traceback (most recent call last):
  File "src/script/ptl-tool.py", line 262, in <module>
    main()
  File "src/script/ptl-tool.py", line 259, in main
    return build_branch(args)
  File "src/script/ptl-tool.py", line 110, in build_branch
    G.remotes.upstream.fetch()
  File "/usr/local/lib/python2.7/dist-packages/git/util.py", line 894, in __getattr__
    return list.__getattribute__(self, attr)
AttributeError: 'IterableList' object has no attribute 'upstream'
```

[the remark this answers is not in the document]
Comment (joscollin) [reply src/script/ptl-tool.py]: When I fix my remote name, it gives the following error, changing the local branch to '* (detached from upstream/master)'
```
jcollin@smithi042:~/test$ env PTL_TOOL_BASE_PATH=refs/remotes/upstream/ src/script/ptl-tool.py --base master 17968
Detaching HEAD onto base: master
Merging PR #17968
Traceback (most recent call last):
  File "src/script/ptl-tool.py", line 262, in <module>
    
  File "src/script/ptl-tool.py", line 259, in main
    
  File "src/script/ptl-tool.py", line 129, in build_branch
    repo.index.add([".githubmap"])
IndexError: list index out of range
```

File: src/script/ptl-tool.py [hunk 2]
```diff
@@ -1,27 +1,84 @@
...
+
 # TODO
+# Fetch PRs by label.
 # Look for check failures?
 # redmine issue update: http://www.redmine.org/projects/redmine/wiki/Rest_Issues
 
+import argparse
 import datetime
 import getpass
 import git
```
Comment (joscollin) [inline src/script/ptl-tool.py]: $ pip install gitpython -> add it in install-deps.sh ?

File: src/script/ptl-tool.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (batrick) [reply src/script/ptl-tool.py]: > src/scripts/ptl-tool.py incorrect path

You probably didn't checkout the PR? It isn't in your Ceph checkout?

[the remark this answers is not in the document]
Comment (batrick) [reply src/script/ptl-tool.py]: > The script is always expecting a remote named 'upstream', when my remote is 'origin git@github.com:ceph/ceph'

You need to use whatever your local ref for upstream is. Perhaps: `refs/remotes/origin/` ?

File: src/script/ptl-tool.py (hunk 2 above)
Comment (batrick) [reply src/script/ptl-tool.py]: No, this is something you should install manually. We don't want that installed for users of Ceph.

File: src/script/ptl-tool.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (joscollin) [reply src/script/ptl-tool.py]: @batrick 
Yes, I have used my local ref for upstream too. `$ env PTL_TOOL_BASE_PATH=refs/remotes/origin/ src/script/ptl-tool.py --base master 17968`. 

But still got the same error:
`AttributeError: 'IterableList' object has no attribute 'upstream'`

File: src/script/ptl-tool.py (hunk 2 above)
Comment (joscollin) [reply src/script/ptl-tool.py]: Oh yes, Sorry my bad :-)

File: src/script/ptl-tool.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (batrick) [reply src/script/ptl-tool.py]: Oh, I see the problem. Working on it...

[the remark this answers is not in the document]
Comment (batrick) [reply src/script/ptl-tool.py]: @joscollin please try again. You will also need to set `PTL_TOOL_BASE_REMOTE`:

PTL_TOOL_BASE_REMOTE=origin

File: src/script/ptl-tool.py [hunk 3]
```diff
@@ -42,51 +104,62 @@
...
+        log.info("Detaching HEAD onto base: {}".format(base))
+        try:
+            base_path = args.base_path + base
+            base = filter(lambda r: r.path == base_path, G.refs)[0]
+        except IndexError:
+            log.error("Branch base does not exist!")
+            sys.exit(1)
 
         # So we know that we're not on an old test branch, detach HEAD onto ref:
         base.checkout()
```
Comment (joscollin) [inline src/script/ptl-tool.py]: Please review https://github.com/batrick/ceph/pull/3 and provide your comments. This works for me. Otherwise, it leaves "detached from origin/master".

File: src/script/ptl-tool.py [hunk 4]
```diff
@@ -42,51 +104,62 @@
...
-        else:
-            base = BASE % "master"
-        log.info("Branch base on {}".format(base))
-        base = filter(lambda r: r.path == base, repo.refs)[0]
+        log.info("Detaching HEAD onto base: {}".format(base))
+        try:
+            base_path = args.base_path + base
+            base = filter(lambda r: r.path == base_path, G.refs)[0]
+        except IndexError:
+            log.error("Branch base does not exist!")
```
Comment (joscollin) [inline src/script/ptl-tool.py]: Please pull https://github.com/batrick/ceph/pull/2

File: src/script/ptl-tool.py [hunk 5]
```diff
@@ -120,49 +193,81 @@ def build_branch(branch_name, pull_requests):
 
         if new_new_contributors:
```
[the remark this answers is not in the document]
Comment (batrick) [reply src/script/ptl-tool.py]: > We are going to run this script in a test machine, mostly in smithi, without the intention to update the contributors list.

Idea is if we have a reviewer we haven't seen before, then the merger should figure out who they are (if possible, skip using EOF/CTRL+D if not) and add them.

>I think it will make .githubmap a mess, as any text is acceptable here.

We can add warnings but I don't really think this is a big deal.

[the remark this answers is not in the document]
Comment (joscollin) [reply src/script/ptl-tool.py]: @batrick 
Agreed.

> Need name for contributor "pagupta"; Reviewed-by:
I was wondering, if it is possible to `.get` the Name and Email ID from github. Let me check if that is possible.

[the remark this answers is not in the document]
Comment (joscollin) [reply src/script/ptl-tool.py]: I don't think it is possible to implement it correctly, as a user can choose not to specify his email address as a Public Email. 

The user can also make his email address private:
"
Keep my email address private
We'll remove your public profile email and use 24334514+joscollin@users.noreply.github.com when performing web-based Git operations and sending email on your behalf. 
"

However, I have sent a query to github support to confirm this.

File: src/script/ptl-tool.py [hunk 6]
```diff
@@ -42,51 +104,62 @@
         m = patt.match(line)
         CONTRIBUTORS[m.group(1)] = m.group(2)
 
-def build_branch(branch_name, pull_requests):
-    repo = git.Repo(".")
+def build_branch(args):
+    base = args.base
+    branch = args.branch
+    label = args.label
```
[the remark this answers is not in the document]
Comment (joscollin) [reply src/script/ptl-tool.py]: @liewegas Could you please provide your inputs on [our discussion here](https://github.com/batrick/ceph/pull/4)  ? Thanks.

@liewegas @yuriw At the moment, the script validates the labels of type `wip-<name>-testing` and not `wip-<name>-testing2`. So would you like to update your labels to say `wip-sage2-testing`, to match the correct format ? (I think that is better). Otherwise, we have to incorporate that trailing number too. Please let me know.

**Comment by batrick** [comment]
@joscollin This is getting a little jumbled with all the merges. How about we merge this and you can write some new PRs against what's in master?

**Comment by joscollin** [comment]
@batrick That's fine. But before that could you please merge https://github.com/batrick/ceph/pull/4, if you agree with those changes, so that it avoids a redo on my side ? I'm approving this PR now, so that you can proceed with merge.

**Comment by batrick** [comment]
Thanks, @joscollin. I appreciate the changes you're making/recommending.

**Comment by joscollin** [comment]
yw @batrick

**Comment by tchaikov** [comment]
@joscollin @batrick please avoid introducing merge commits in PRs if possible, thanks.

**Comment by joscollin** [comment]
@tchaikov Yes, that's another reason we have merged this and will continue the work by creating new PRs.

**Comment by tchaikov** [comment]
@joscollin i don't get you. let me rephrase my suggestion: please squash related commits into a single one before merging them.

**Comment by joscollin** [comment]
@tchaikov Oh I see your point now. I was submitting the PRs (4 PRs created) to  `batrick:ptl-tool-2` and batrick was merging it for me. 3 PRs merged. :-)
